src/server/pfs/server/driver_file.go

```diff
@@ -356,17 +362,69 @@ func (d *driver) walkFile(ctx context.Context, file *pfs.File, cb func(*pfs.File
 			})
 		}),
 	}
+	if paginationMarker != nil {
+		pathRange := &pfs.PathRange{}
+		if reverse {
+			pathRange.Upper = paginationMarker.Path
+		} else {
+			pathRange.Lower = paginationMarker.Path
+		}
+		opts = append(opts, WithPathRange(pathRange))
+	}
 	s := NewSource(commitInfo, fs, opts...)
 	s = NewErrOnEmpty(s, newFileNotFound(commitInfo.Commit.ID, p))
+	if reverse {
+		return d.walkFileReverse(ctx, s, number, cb)
+	}
 	err = s.Iterate(ctx, func(fi *pfs.FileInfo, f fileset.File) error {
+		if number == 0 {
+			return errutil.ErrBreak
+		}
+		number--
 		return cb(fi)
 	})
-	if p == "" && pacherr.IsNotExist(err) {
+	if (p == "" && pacherr.IsNotExist(err)) || err == errutil.ErrBreak {
 		err = nil
 	}
 	return err
 }
 
+func (d *driver) walkFileReverse(ctx context.Context, s Source, number int64, cb func(*pfs.FileInfo) error) error {
```

Comment:
We should reuse `listFileReverse` since it is the same code. Maybe rename it to `reverseFiles`. It would just be a utility function for reversing a certain number of files from a source.

Reply:
I've moved the circular buffer into a struct with helper functions which removes a lot of the re-used code.